pandora.py

```diff
@@ -9,30 +9,65 @@ class AudioFormat(enum.Enum):
     AACPLUS = 'aacplus'
 
 
+class Art:
+    def __init__(self, art: dict) -> None:
+        self.x_small = art.get('90')
+        self.small = art.get('130')
+        self.medium = art.get('500')
+        self.large = art.get('640')
+        self.x_large = art.get('1080')
+
+
 class Track:
     def __init__(self, data: dict) -> None:
+        # Must have
         self.title = data['songTitle']
+        self.album_title = data['albumTitle']
+        self.artist_name = data['artistName']
         self.music_id = data['musicId']
+        self.artist_music_id = data['artistMusicId']
+        self.pandora_id = data['pandoraId']
         self.token = data['trackToken']
-        self.length = data['trackLength']
-        self.gain = float(data['fileGain'])
+        self.station_id = data['stationId']
         self.audio_url = data['audioURL']
-        self.audio_encoding = data['audioEncoding']
-        self.art = {i['size']: i['url'] for i in data['albumArt']}
+        # Nice to have
+        self.rating = int(data.get('rating', 0))
+        self.length = int(data.get('trackLength', 0))
+        self.gain = float(data.get('fileGain', 0.0))
+        self.audio_encoding = data.get('audioEncoding', '')
+        self.art = Art({i['size']: i['url'] for i in data.get('albumArt', [])})
```

Comment:
I find it a bit odd to have one `Art` object containing multiple sizes, wouldn't a list of `Art` containing the size/url make the most sense?

Reply:
No, this way you can do track.art.x_large.

Reply:
Since the objects are mostly prop dumpsters what do think about make pretty much everything actual ```@property``'s?

Reply:
GObject properties eventually probably, otherwise not too valuable.

Reply:
> GObject properties eventually probably, otherwise not too valuable.

Even the value of that questionable. I mean one the most obvious advantage of GObjects is being able to connect to the prop's notify signal and there isn't much value in that if the prop value never change?

Reply:
Some properties can change, if we hook up that machinery, like station names, like/ban status of songs, etc.

Reply:
> Some properties can change, if we hook up that machinery, like station names and like/ban status of songs, etc.

Very true. If you like I can make the objects GObjects now pretty easily?

Reply:
Eh, we can just flip that switch once somebody implements logic that actually has dynamic properties. Feel free to keep extending the API towards that though =)

Reply:
> Feel free to keep extending the API coverage towards that though =)

I will. I get a week off for Christmas. So the plan was to really hunker down a peep the API by systematically going step by step feature by feature to see how things are done after I figure out all of the methods. It's not much good to have all of the methods and not know how to dance,lol!!!

Reply:
I also seem to be running into rate limiting and/or Pandora expects that the authToken be reused because I get:

```soupy.SoupException: STREAM_VIOLATION; TOO_MANY_REQUESTS (429): STREAM_VIOLATION```

And

```soupy.SoupException: AUTH_INVALID_TOKEN; BAD_REQUEST (400): Invalid auth token```

If I run the dummy code in __main_.py more than once in like 5 mins.

Reply:
Hmm, I've certainly ran it maybe a dozen times within 5 mins. Should check if the website reuses the authtoken.

Reply:
The problem only seems to exist for paid accounts. None of my free test accounts have a problem, but as soon as I switch to my paid account the error is back...

Reply:
Well you still need an email but it will reuse the autToken the method arg name is existingAuthToken and the password was left empty.

Reply:
> The problem only seems to exist for paid accounts.

That sounds backwards :P

Reply:
It really does...

Reply:
Ugh, looks like the vrest website free trial expires soon.

Reply:
So just start a new trial under a new email...

Reply:
*effort*

Reply:
Should just learn machine learning to a degree where it can automatically implement clients for a specified API. Will save so much time!

Reply:
You'd still need to document the methods and provide context. Machines are not good at context...